### PR TITLE
refactor(ecocredit/marketplace): BuyDirect

### PR DESCRIPTION
**Pull request by technicallyty**
## Description

- refactors Buy to only deal with filtered buy orders
- creates new BuyDirect rpc method for purchasing directly from sell orders.
- fixes an issue with calculating the total cost
- adds test case for fractional purchase

Closes: #905 

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules)
- [ ] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed 
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

**Review by robert-zaremba — COMMENTED** [review]
`int64` roundtrip doesn't look good. Let's revisit.

**Review by robert-zaremba — COMMENTED** [review]
`int64` roundtrip doesn't look good. Let's revisit.

File: x/ecocredit/server/marketplace/buy.go
Comment (technicallyty) [reply x/ecocredit/server/marketplace/buy.go]: switched to string https://github.com/regen-network/regen-ledger/pull/967/commits/1462953d9e8f8d93e7e5be2aaa88f7b88f3a5cd3

File: x/ecocredit/server/marketplace/buy.go
Comment (technicallyty) [reply x/ecocredit/server/marketplace/buy.go]: switched to string https://github.com/regen-network/regen-ledger/pull/967/commits/1462953d9e8f8d93e7e5be2aaa88f7b88f3a5cd3

File: x/ecocredit/server/marketplace/buy.go
Comment (technicallyty) [reply x/ecocredit/server/marketplace/buy.go]: it's to eliminate anything to the right of the decimal. otherwise, cannot convert to `sdk.Int`

File: x/ecocredit/server/marketplace/buy_direct.go
Comment (technicallyty) [reply x/ecocredit/server/marketplace/buy_direct.go:55]: i don't think it is, theres nothing from the request in these lines. sellOrder is pulled from state. i think if anything it would be some sort of internal/server error

File: proto/regen/ecocredit/marketplace/v1/tx.proto
Comment (ryanchristo) [inline proto/regen/ecocredit/marketplace/v1/tx.proto:29]:
```suggestion
  // BuyDirect purchases credits directly from the specified sell order.
  rpc BuyDirect(MsgBuyDirect) returns (MsgBuyDirectResponse);
```

File: proto/regen/ecocredit/marketplace/v1/tx.proto
Comment (ryanchristo) [inline proto/regen/ecocredit/marketplace/v1/tx.proto:193]:
```suggestion
  // sell_order_id is the sell order ID against which the buyer is trying
  // to buy.
```

File: proto/regen/ecocredit/marketplace/v1/tx.proto
Comment (ryanchristo) [inline proto/regen/ecocredit/marketplace/v1/tx.proto:200]: I think we might actually want to keep this as `bid_price`. The `ask_price` and `bid_price` in both sell and buy orders are also price per credit and changing the name here might cause confusion.

File: x/ecocredit/marketplace/msg_buy_direct.go
Comment (ryanchristo) [reply x/ecocredit/marketplace/msg_buy_direct.go:33]: I don't think a separate helper function is necessary. We would only be able to extract three of these checks (quantity, retirement location, and bid price). I don't have a strong preference though.

File: x/ecocredit/marketplace/msg_buy_direct_test.go
Comment (ryanchristo) [inline x/ecocredit/marketplace/msg_buy_direct_test.go:11]: :pray:

File: x/ecocredit/server/marketplace/buy.go
Comment (ryanchristo) [inline x/ecocredit/server/marketplace/buy.go:19]: Both direct and filtered buy orders will still be used alongside the orderbook as far as I understand.

```suggestion
// Buy creates a buy order that matches sell orders based on filter criteria.
```

File: x/ecocredit/server/marketplace/buy.go
Comment (ryanchristo) [inline x/ecocredit/server/marketplace/buy.go:21]: Should we do this in `ValidateBasic` rather than letting the message fail in the mempool?

File: x/ecocredit/server/marketplace/buy.go
Comment (ryanchristo) [inline x/ecocredit/server/marketplace/buy.go:27]:
```suggestion
```

**Review by ryanchristo — COMMENTED** [review]
Looking good. Should we use `bid_price` rather than changing the name? Do we want to support multiple orders in the same message as we do for `Sell` and `Buy`?

File: x/ecocredit/server/marketplace/buy.go
Comment (technicallyty) [reply x/ecocredit/server/marketplace/buy.go:21]: ah yeah good idea - fixed https://github.com/regen-network/regen-ledger/pull/967/commits/5f8029c324e18051550fbea5bb1360be95620ada

File: x/ecocredit/marketplace/msg_buy_direct_test.go
Comment (technicallyty) [reply x/ecocredit/marketplace/msg_buy_direct_test.go]: https://github.com/regen-network/regen-ledger/pull/967/commits/e1a327e867dba282cc4cbba7bdd755d64af3183f

File: proto/regen/ecocredit/marketplace/v1/tx.proto
Comment (technicallyty) [reply proto/regen/ecocredit/marketplace/v1/tx.proto:200]: i just wasn't sure `bid_price` made much sense in this context, as BuyDirect doesn't really involve bidding on anything. i was thinking of it as similar to going to the grocery store, I wouldn't really "bid" on a bag of chips 🤷🏻

Comment (technicallyty) [reply proto/regen/ecocredit/marketplace/v1/tx.proto]: ok, switched back https://github.com/regen-network/regen-ledger/pull/967/commits/e1529683ffd0951557abd4d50f1471b4b4acfa73

File: x/ecocredit/server/marketplace/buy.go
Comment (technicallyty) [reply x/ecocredit/server/marketplace/buy.go]: ok created an orderOptions struct to slim down the function

Comment (technicallyty) [reply x/ecocredit/server/marketplace/buy.go]: https://github.com/regen-network/regen-ledger/pull/967/commits/2e27c649f7a0de1c7fec9929ec6b26e3df21353d

File: proto/regen/ecocredit/marketplace/v1/tx.proto
Comment (ryanchristo) [inline proto/regen/ecocredit/marketplace/v1/tx.proto:212]: We use `repeated Order` in `Buy` to support multiple buy orders. I think we should preserve that here. Imagine a user browsing the marketplace, selecting multiple sell orders (i.e. "add to cart"), and then purchasing.

**Review by ryanchristo — COMMENTED** [review]
I think we should support multiple buy orders. Otherwise looks good.

**Comment by ryanchristo** [comment]
We'll also need a followup task for updating the CLI command.

File: proto/regen/ecocredit/marketplace/v1/tx.proto
Comment (technicallyty) [reply proto/regen/ecocredit/marketplace/v1/tx.proto]: added: https://github.com/regen-network/regen-ledger/pull/967/commits/fe36dd2c3c8ef771c5df022bcc90f181cbe82cf2

File: x/ecocredit/server/marketplace/buy.go
Comment (robert-zaremba) [reply x/ecocredit/server/marketplace/buy.go]: hmm, I thought that you want to convert `{1.001, exponent = 4}` to `Int(10010)`. Why cutting the decimal places?

Comment (technicallyty) [reply x/ecocredit/server/marketplace/buy.go]: so the bulk of that is already done in the lines above the reviewed snippet:

```go
	amountDec, err := amtCredits.Mul(multiplier)
	if err != nil {
		return sdk.Int{}, err
	}
```

however, in cases of small costs, you can end up with remaining decimals, which need to be eliminated, otherwise you can't put it in `sdk.Int`. 

example:
`quantity of credits to purchase: 10.54321`
`Price per credit: 30uatom`

`Total Cost: 316.2963`

i dont think this would happen too often as 30uatom is a suuuuuuper tiny amt to pay for credits but i guess it could happen with user input error. 

only question left is what to do in these cases: trim decimal? error out? something else?

Comment (robert-zaremba) [reply x/ecocredit/server/marketplace/buy.go]: I see. Well, there is no smaller unit than `uatom`. So you can't pay `0.1` uatom. We should either trim decimal or round it. I prefer trimming.

Could you add comments next to the code?

Is there any other way to cut decimals? It feels to me that the `Quo` is suboptimal. Maybe we can create a new task to look at it to not block this one?

File: x/ecocredit/server/marketplace/buy_direct.go
Comment (robert-zaremba) [reply x/ecocredit/server/marketplace/buy_direct.go]: OK

File: proto/regen/ecocredit/marketplace/v1/tx.proto
Comment (aaronc) [inline proto/regen/ecocredit/marketplace/v1/tx.proto:26]: Since this isn't support it yet, we should just delete `Buy` and add it back later

File: x/ecocredit/marketplace/msg_buy_test.go
Comment (aaronc) [inline x/ecocredit/marketplace/msg_buy_test.go:9]: why not just delete all this?

File: x/ecocredit/marketplace/msg_buy_test.go
Comment (robert-zaremba) [inline x/ecocredit/marketplace/msg_buy_test.go:137]: please use `/* .... */` comment to avoid unnecessary changes / long diff.

**Review by aaronc — CHANGES_REQUESTED** [review]
Please delete the `Buy` method and also `buy.go` so the diff shows `buy_direct.go` renamed to `buy.go`. That will be easier to review

File: x/ecocredit/server/marketplace/utils.go
Comment (robert-zaremba) [inline x/ecocredit/server/marketplace/utils.go:196]: please add TODO to update this code after https://github.com/regen-network/regen-ledger/issues/987

File: x/ecocredit/server/marketplace/utils.go
Comment (robert-zaremba) [inline x/ecocredit/server/marketplace/utils.go:186]:
```suggestion
	unitPrice, err := math.NewPositiveFixedDecFromString(pricePerCredit.String(), amtCredits.NumDecimalPlaces())
```

File: x/ecocredit/server/marketplace/utils.go
Comment (robert-zaremba) [inline x/ecocredit/server/marketplace/utils.go:190]:
```suggestion
	cost, err := amtCredits.Mul(multiplier)
```

File: x/ecocredit/server/marketplace/utils.go
Comment (robert-zaremba) [inline x/ecocredit/server/marketplace/utils.go:203]:
```suggestion
	costIntStr := amountDec.String()
	costInt, ok := sdk.NewIntFromString(amtNeededStr)
```

**Review by robert-zaremba — APPROVED** [review]
pre-approving. Left few suggestions.

File: proto/regen/ecocredit/marketplace/v1/tx.proto
Comment (technicallyty) [reply proto/regen/ecocredit/marketplace/v1/tx.proto]: deleted

File: x/ecocredit/marketplace/msg_buy_test.go
Comment (technicallyty) [reply x/ecocredit/marketplace/msg_buy_test.go]: deleted

File: x/ecocredit/marketplace/msg_buy_test.go
Comment (technicallyty) [reply x/ecocredit/marketplace/msg_buy_test.go]: deleted per https://github.com/regen-network/regen-ledger/pull/967#discussion_r844401876

File: x/ecocredit/server/marketplace/utils.go
Comment (technicallyty) [reply x/ecocredit/server/marketplace/utils.go]: added

File: x/ecocredit/server/marketplace/utils.go
Comment (robert-zaremba) [inline x/ecocredit/server/marketplace/utils.go:204]: I just found that we have `BigInt` function - so instead of doing a roundtrip through string, let's use `big.Int` -> `sdk.Int`

Comment (technicallyty) [reply x/ecocredit/server/marketplace/utils.go]: https://github.com/regen-network/regen-ledger/pull/967/commits/ae9ef5995108e1a7484dcfbaf0a147cd8ff2865e

File: proto/regen/ecocredit/marketplace/v1/tx.proto
Comment (aaronc) [inline proto/regen/ecocredit/marketplace/v1/tx.proto:26]: `MsgBuy`, etc. should also get deleted

Comment (technicallyty) [reply proto/regen/ecocredit/marketplace/v1/tx.proto:26]: okay removed everything related to MsgBuy

File: x/ecocredit/marketplace/msg_buy_direct_test.go
Comment (ryanchristo) [inline x/ecocredit/marketplace/msg_buy_direct_test.go:44]: Might be nice to consistently give each field its own line as you have done in the first test case  to improve readability. Here and below, even if only one field, but ok to leave as is.

File: x/ecocredit/marketplace/msg_buy_direct_test.go
Comment (ryanchristo) [inline x/ecocredit/marketplace/msg_buy_direct_test.go:97]: Might be nice to provide the whole error and use equals rather than contains but ok leave as is.

File: x/ecocredit/server/marketplace/buy_direct.go
Comment (ryanchristo) [inline x/ecocredit/server/marketplace/buy_direct.go:50]:
```suggestion
			return nil, sdkerrors.ErrInvalidRequest.Wrapf("bid price denom does not match ask price denom: "+
```

File: x/ecocredit/server/marketplace/buy_direct.go
Comment (ryanchristo) [inline x/ecocredit/server/marketplace/buy_direct.go:65]:
```suggestion
		sellOrderAskPrice, ok := sdk.NewIntFromString(sellOrder.AskPrice)
		if !ok {
			return nil, fmt.Errorf("could not convert %s to %T", sellOrder.AskPrice, sdk.Int{})
		}
		sellOrderPriceCoin := sdk.Coin{Denom: market.BankDenom, Amount: sellOrderAskPrice}
		if sellOrderAskPrice.GT(order.BidPrice.Amount) {
			return nil, sdkerrors.ErrInvalidRequest.Wrapf("price per credit too low: sell order ask per credit: %v, request: %v", sellOrderPriceCoin, order.BidPrice)
		}

		// check address has the total cost (price per * order quantity)
		bal := k.bankKeeper.GetBalance(sdkCtx, buyerAcc, order.BidPrice.Denom)
		cost, err := getTotalCost(sellOrderAskPrice, creditOrderQty)
```

File: x/ecocredit/server/marketplace/utils.go
Comment (ryanchristo) [inline x/ecocredit/server/marketplace/utils.go:40]:
```suggestion
// fillOrder moves credits and coins according to the order. It will:
// - update a sell order, removing it if quantity becomes 0 as a result of this purchase.
// - remove the purchaseQty from the seller's escrowed balance.
// - add credits to the buyer's tradable/retired address (based on the DisableAutoRetire field).
// - update the supply accordingly.
// - send the coins specified in the bid to the seller.
```

File: x/ecocredit/server/marketplace/utils.go
Comment (ryanchristo) [inline x/ecocredit/server/marketplace/utils.go:139]:
```suggestion
			RetiredAmount:  "0",
```

File: x/ecocredit/server/marketplace/utils.go
Comment (ryanchristo) [inline x/ecocredit/server/marketplace/utils.go:199]: Tagging #987 and #988 to make sure we add it there if this gets merged first.

**Review by ryanchristo — APPROVED** [review]
Pre-approving with a few minor suggestions.

File: x/ecocredit/marketplace/msg_buy_direct_test.go
Comment (technicallyty) [reply x/ecocredit/marketplace/msg_buy_direct_test.go]: expanded them 👍🏻 https://github.com/regen-network/regen-ledger/pull/967/commits/df038d4865e2beaabcb5a7ba0a310216b9e84535

**Pull request merged**
